cmd/query/app/server_test.go

```diff
@@ -53,10 +62,314 @@ func TestCreateTLSServerError(t *testing.T) {
+		expectError      bool
+		expectServerFail bool
+	}{
+		// {
+		// 	name: "",
+		// 	serverTLS: tlscfg.Options{
+		// 		Enabled:      true,
+		// 		CertPath:     "invalid/path",
+		// 		KeyPath:      "invalid/path",
+		// 		ClientCAPath: "invalid/path",
+		// 	},
+		// 	clientTLS: tlscfg.Options{
+		// 		Enabled:    true,
+		// 		CertPath:   "invalid/path",
+		// 		KeyPath:    "invalid/path",
+		// 		CAPath:     "invalid/path",
+		// 		ServerName: "example.com",
+		// 	},
+		// 	expectError:      false,
+		// 	expectServerFail: false,
+		// },
+		{
+			name: "Should pass with insecure connection",
+			serverTLS: tlscfg.Options{
+				Enabled: false,
+			},
+			clientTLS: tlscfg.Options{
+				Enabled: false,
+			},
+			expectError:      false,
+			expectServerFail: false,
+		},
+		{
+			name: "should fail with TLS client to untrusted TLS server",
+			serverTLS: tlscfg.Options{
+				Enabled:  true,
+				CertPath: testCertKeyLocation + "/example-server-cert.pem",
+				KeyPath:  testCertKeyLocation + "/example-server-key.pem",
+			},
+			clientTLS: tlscfg.Options{
+				Enabled:    true,
+				ServerName: "example.com",
+			},
+			expectError:      true,
+			expectServerFail: false,
+		},
+		{
+			name: "should fail with TLS client to trusted TLS server with incorrect hostname",
+			serverTLS: tlscfg.Options{
+				Enabled:  true,
+				CertPath: testCertKeyLocation + "/example-server-cert.pem",
+				KeyPath:  testCertKeyLocation + "/example-server-key.pem",
+			},
+			clientTLS: tlscfg.Options{
+				Enabled:    true,
+				CAPath:     testCertKeyLocation + "/example-CA-cert.pem",
+				ServerName: "nonEmpty",
+			},
+			expectError:      true,
+			expectServerFail: false,
+		},
+		{
+			name: "should pass with TLS client to trusted TLS server with correct hostname",
+			serverTLS: tlscfg.Options{
+				Enabled:  true,
+				CertPath: testCertKeyLocation + "/example-server-cert.pem",
+				KeyPath:  testCertKeyLocation + "/example-server-key.pem",
+			},
+			clientTLS: tlscfg.Options{
+				Enabled:    true,
+				CAPath:     testCertKeyLocation + "/example-CA-cert.pem",
+				ServerName: "example.com",
+			},
+			expectError:      false,
+			expectServerFail: false,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			serverOptions := &QueryOptions{TLSHTTP: test.serverTLS}
+			httpServer, err := createHTTPServer(&querysvc.QueryService{}, serverOptions, opentracing.NoopTracer{}, zap.NewNop())
+
+			if test.expectServerFail {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.NotNil(t, httpServer)
+			}
+			httpListener, err := net.Listen("tcp", "localhost:"+fmt.Sprintf("%d", ports.QueryHTTP))
+			if test.expectServerFail {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+			if serverOptions.TLSHTTP.Enabled {
+
+				tlsCfg, _ := serverOptions.TLSHTTP.Config()
+				tlsCfg.Rand = rand.Reader
+				tlsHTTPListener := tls.NewListener(httpListener, tlsCfg)
```

Comment:
this should use the actual business logic, to test it, not replicate it here

Reply:
Sorry,  I dont fully understand what you mean.  

This was the relevant minimum functionality / logic that i derive form the tesing of HTTP Server.  I need to do this  because mTLS with a bad certificate isn't verifyable with `TLS.Dial` (i.e.  no error is given. The error is given in the first read only.   [This](https://github.com/golang/go/issues/28779#issuecomment-442694636) could shed some light on the issue.  So in order to read something that would be accepted with no errors,  I resorted to the logic defined in https://github.com/jaegertracing/jaeger/blob/master/cmd/query/app/http_handler_test.go .  

Could you please elaborate a little bit on the error and what is expected?  thanks.

Reply:
I mean that it is largely repeating what `func (s *Server) serveHTTP(httpListener net.Listener)` function does, which is production code, and we should try to use that rather than repeating it in the test.

Reply:
Aah.  okay.  I wrote this method to test `func createHTTPServer(querySvc *querysvc.QueryService, queryOpts *QueryOptions, tracer opentracing.Tracer, logger *zap.Logger) (*http.Server, error)`  function without the `Server`.  Once #2338 is resolved,  only then i can use `Server` directly with TLS.  Ill get right on it.